src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs

```diff
@@ -2523,6 +2521,59 @@ private bool Fix(int iParam, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
             }
         }
 
+        /// <summary>
+        /// Takes the dynamic flags from both types and applies them onto the target.
+        /// </summary>
+        internal static TypeSymbol MergeDynamic(TypeSymbol first, TypeSymbol second, TypeSymbol target, Compilation compilation)
+        {
+            // SPEC: 4.7 The Dynamic Type
+            //       Type inference (7.5.2) will prefer dynamic over object if both are candidates.
+            if (first.Equals(second, TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds | TypeCompareKind.IgnoreDynamic))
+            {
+                return target;
+            }
+            ImmutableArray<bool> flags1 = CSharpCompilation.DynamicTransformsEncoder.EncodeWithoutCustomModifierFlags(first, RefKind.None);
+            ImmutableArray<bool> flags2 = CSharpCompilation.DynamicTransformsEncoder.EncodeWithoutCustomModifierFlags(second, RefKind.None);
+            Debug.Assert(flags1.Length == flags2.Length);
+            ImmutableArray<bool> mergedFlags = flags1.SelectAsArray((f, index, other) => f | other[index], flags2);
```

Comment:
Consider adding a `static ImmutableArray<TResult> ZipAsArray<T1, T2, TResult>(this ImmutableArray<T1> a, ImmutableArray<T2> b, Func<T1, T2, TResult> f)` extension method. (There are a few calls to `System.Linq.Enumerable.Zip` in other parts of the code that could be changed to use `ZipAsArray` as well.)